src/operator/tensor/matrix_op-inl.h

```diff
@@ -1583,6 +1527,91 @@ void TileOpBackward(const nnvm::NodeAttrs& attrs,
       attrs, ctx, newInputs, req, newOutputs, rshapes.first);
 }
 
+struct ReverseParam : public dmlc::Parameter<ReverseParam> {
+  nnvm::Tuple<int> axis;
+  DMLC_DECLARE_PARAMETER(ReverseParam) {
+    DMLC_DECLARE_FIELD(axis)
+    .describe("The axis which to reverse elements.");
+  }
+};
+
+/**
+* \brief Calculate Tensor dim offset
+* \param shape
+* \param offset
+*/
+inline void ReverseCalculateOffset(const TShape& shape, std::vector<index_t> &offset) { // NOLINT(*)
+  int ndim = shape.ndim();
+  offset.resize(ndim);
+  offset[ndim - 1] = 1;
+  for (int i = ndim - 2; i >= 0; --i) {
+    offset[i] = offset[i + 1] * shape[i + 1];
+  }
+}
+
+
+template<typename xpu>
+void ReverseOpForward(const nnvm::NodeAttrs& attrs,
+                      const OpContext& ctx,
+                      const std::vector<TBlob>& inputs,
+                      const std::vector<OpReqType>& req,
+                      const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  const ReverseParam& param = nnvm::get<ReverseParam>(attrs.parsed);
+  CHECK_EQ(inputs[0].type_flag_, outputs[0].type_flag_);
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+
+  const TShape& ishape = inputs[0].shape_;
+  std::vector<index_t> offset;
+  ReverseCalculateOffset(ishape, offset);
+
+  std::vector<index_t> dims(ishape.ndim());
+  std::copy(ishape.begin(), ishape.end(), dims.begin());
+  std::vector<uint8_t> reverse_dims(ishape.ndim(), 0);
+  for (auto axis_iter = param.axis.begin() ; axis_iter!= param.axis.end(); ++axis_iter) {
+    CHECK_LT(*axis_iter, ishape.ndim());
+    reverse_dims[*axis_iter] = 1;
+  }
+
+#ifdef __CUDACC__
+  mshadow::Tensor<xpu, 1, uint8_t> workspace =
+    ctx.requested[0].get_space_typed<xpu, 1, uint8_t>(
+      mshadow::Shape1(ishape.ndim() * sizeof(index_t) * 3), s);
+
+  auto dims_workspace = workspace.dptr_;
+  auto reverse_dims_workspace = workspace.dptr_ + ishape.ndim() * sizeof(index_t);
+  auto offset_workspace = workspace.dptr_ + ishape.ndim() * sizeof(index_t) * 2;
+
+  cudaMemcpyAsync(dims_workspace, thrust::raw_pointer_cast(dims.data()),
```

Comment:
You can pass the shapes as arguemnts like Map(Shape<3> a, ...) no need to copy here